design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,178 @@
+
+#### Flow 2: Optimized Run
+
+Optimized run uses the previous GC run output, to perform a partial scan of the branch to remove uncommitted garbage.
+
+##### Step 1. Analyze Data and Perform Cleanup for old entries (GC client)
+
+1. Read previous run's information
+   1. Previous `Uncommitted DF`
+   2. Last read slice
+   3. Last run's timestamp
+2. Listing of lakeFS repository uncommitted objects  
+    See previous for steps
+3. Listing of lakeFS repository committed objects (optimized)
+   1. Read addresses from branch's new commits (all new commits down to the last GC run timestamp) -> `Committed DF`
+4. Find candidates for deletion
+   1. Subtract `Committed DF` from previous run's `Uncommitted DF`
+   2. Subtract current `Uncommitted DF` from previous run's `Uncommitted DF`
+   3. The result is a list of files that can be safely removed
+
+>**Note:** This step handles cases of objects that were uncommitted during previous GC run and are now deleted
+
+##### Step 2. Analyze Data and Perform Cleanup for new entries (GC client)
+
+1. Listing namespace objects (optimized)
+   1. Read all objects directly from object store
+   2. Skip slices that are newer than < TOKEN_EXPIRY_TIME >
+   3. Using the slices, stop after reading the last slice read by previous GC run -> `Store DF`
+2. Find candidates for deletion
+   1. Subtract `Committed DF` from `Store DF`
+   2. Subtract current `Uncommitted DF` from `Store DF`
+   3. Filter files in special paths
+   4. The remainder is a list of files which can be safely removed
+3. Save run data (see: [GC Saved Information](#gc-saved-information))
+
+### GC Saved Information
+
+For each GC run, save the following information using the GC run id as detailed in this [proposal](https://github.com/treeverse/cloud-controlplane/blob/main/design/accepted/gc-with-run-id.md):
+1. Save metaranges in `_lakefs/gc/run_id/metadata/`
+2. Save `Uncommitted DF` in `_lakefs/gc/run_id/uncommitted.parquet`
+3. Add the following to the GC report:
+   1. Run start time
+   2. Last read slice
+
+## Limitations
+
+* Since this solution relies on the new repository structure, it is not backwards compatible. Therefore, another solution will be required for existing 
+repositories
+* Even with the given optimizations, the GC process is still very much dependent on the amount of changes that were made on the repository
+since the last GC run.
+
+## Performance Requirements
+
+The heaviest operation during the GC process, is the namespace listing. And while we added the above optimizations to mitigate
+this process, the fact remains - we still need to scan the entire namespace (in the Clean Run mode).
+Performing tests against an AWS S3 bucket using a Databricks notebook on a m4.large cluster, we've observed listing of 
+~500,000 objects takes approximately 25 seconds.
+We can estimate that on a repository with ~1B objects, using ~500K object size slices, and using 10 workers - listing will take
+around 1.5 hours.
```

Comment:
How do we handle failures in such a long run - say, a crash after an hour of execution?
Does the entire process need to be rerun?
Maybe we can parallelize the GC itself, instead of just the listing?

Reply:
GC job will be parallelized, but we can't rely on failed runs. To get a correct picture of the namespace the process must run successfully and without errors.
Relying on partial data, and failed jobs will increase risk of deleting committed data - which is something that absolutely must not happen